README.md
# yandex-recognizer
Node JS interface to yandex speech recognition service

## Questions and Bug Reports
* mailing list: Victor.Vazin@gmail.com

## Installation
Install the yandex-recognizer and it's dependencies by executing
the following `NPM` command.
```
npm install yandex-recognizer --save
```
## Troubleshooting
The yandex-recognizer depends on several other packages. These are.

* node-uuid
* websocket

Ensure, that your user has write permission to wherever the node modules
are being installed.

QuickStart
==========
Add require to your app:
```
var yandex = require('yandex-recognizer');
```
Create service:
```
var service = yandex.Recognizer(config);
```
`config` object must define at least 5 properties:
```
config = {

	onConnect: function(sessionId, code),
	onResult: function(data),
	onError: function(e),
	onClose: function(e),
	apikey: 'YOUR-OWN-YANDEX-API-KEY'

}
```
All other properties will be set by default (see source code for all available).

Connect service to your app:
```
service.connect();
```
`onConnect(sessionId, code)` method will launched in case of successfull connect. Here you send sound data for recognition:
```
service.send(db, cbLength);

// db (typeof Buffer) - your sound file data.
// cbLength (typeof Number) - length of chunk buffer.
```
`onResult(data)` method receive recognized text and some other data.

`onError(e)` method receive error massages.

`onClose(e)` method receive close massage.

## Limitations
Input audio file format may be now only: 

* Sample Encoding: 16-bit Signed Integer PCM, Sample Rate: 16000, 1 Channel.

## Next Steps
 * [yandex-recognizer documentation](https://github.com/AirGraph/yandex-recognizer)
 * [example](https://github.com/AirGraph/yandex-recognizer/tree/master/example)


HISTORY.md
0.3.0 06-18-2016
----------------
* Added `utteranceSilence` advanced option...

0.2.4 05-09-2016
----------------
* Minor patch...

0.2.2 03-24-2016
----------------
* Added `onClose` event handler...

0.2.0 02-13-2016
----------------
* Fixed a problem with partial results for large input PCM file...

0.1.0 02-06-2016
----------------
* Fixed a problem with splitting large input PCM file into chunks...

0.0.1 01-27-2016
----------------
* Initial release support of yandex speech recognition service using native node.js interface


example/test.js
//	Yandex Speech Recognition for Node JS 4.3.
//		Version 0.0.3.
//			Copyright (c) Jungle Software, 2016.

var	commandLineArgs = require('command-line-args'),
	assert = require('assert'),
	fs = require('fs'),

	yandex = require('yandex-recognizer'),

	options = commandLineOptions(),
	outSize, db, service, cbLength,
	resText = '', curText = '';

function main() {
	fs.stat(options.file, function (err, stats){

		assert.equal(null, err);
		outSize = stats.size;

		fs.open(options.file, 'r', function (err, fd){

			assert.equal(null, err);
			db = new Buffer(outSize);

			fs.read(fd, db, 0, outSize, null, function (err, bytesRead, buffer) {

				assert.equal(null, err);
				service = yandex.Recognizer({

					onConnect: onConnect,
					onResult: onResult,
					onClose: onClose,
					onError: onError,
					apikey: 'YOUR-OWN-YANDEX-API-KEY',

				});

				service.connect();

			});
		});
	});
}

function onConnect(sessionId, code) {

	console.log('sessionId: ' + sessionId);
	console.log('code: ' + code + '\n');
	service.send(db, cbLength);

}

function onResult(data) {

	if(data.uttr) resText += data.text;
	else curText = data.text;

}

function onClose(e) {

	console.log('wasClean: ' + e.wasClean);
	console.log('code: ' + e.code);
	console.log('reason: ' + e.reason + '\n');
	console.log('Final recognition: ' + resText + curText);

}

function onError(e) {

	console.log('onError:');
	for(var i in e) { console.log(i + ': ' + e[i]); }

}

function commandLineOptions() {

	var clo = commandLineArgs([

		{ name: "file", alias: "f", type: String },
		{ name: "chunk", alias: "c", type: Number }

	]), options = clo.parse();

	if ("file" in options) {

		if ("chunk" in options) {

			cbLength = options.chunk;
			return options;

		} else {

			cbLength = 32000;
			console.log('Chunk length is set to 32000 by default...\n');
			return options;

		}
	}

	console.log(clo.getUsage({

		title: "Usage",
		description: "node ysr -f|--file name.ext [-c|--chunk value]"

	}));

	process.exit();

}

main();


yandex-recognizer.js
//	Yandex Recognizer for Node JS 4.4.5
//		Version 0.3.0.
//			Copyright (c) Yandex & Jungle Software, 2016.

var W3CWebSocket = require('websocket').w3cwebsocket,
		uuid = require('node-uuid');

(function (namespace) {
	'use strict';

	/** Available audio formats.
	 * @readonly
	 * @enum
	 */
	namespace.FORMAT = { PCM16: 'audio/x-pcm;bit=16;rate=16000' };

	/** Recognizer default config.
	 * @readonly
	 * @enum
	 */
	namespace._defaults = {

		uuid: uuid.v1(),
		format: namespace.FORMAT.PCM16,
		url: 'wss://webasr.yandex.net/asrsocket.ws',
		applicationName: 'jsapi',

		utteranceSilence: 240,
		partialResults: true,
		punctuation: false,
		allowStrongLanguage: true,
		model: 'freeform',
		lang: 'ru-RU'

	};

	/**
	 * Copies all the properties of `config` to the specified `object`.
	 * There are two levels of defaulting supported:
	 *
	 *	_apply(obj, { a: 1 }, { a: 2 });
	 *	//obj.a === 1
	 *
	 *	_apply(obj, {	 }, { a: 2 });
	 *	//obj.a === 2
	 *
	 * @param {Object} object The receiver of the properties.
	 * @param {Object} config The primary source of the properties.
	 * @param {Object} [defaults] The properties default value.
	 * @return {Object} returns `object`.
	 */
	namespace._enumerables = [//'hasOwnProperty', 'isPrototypeOf', 'propertyIsEnumerable',
											 'valueOf', 'toLocaleString', 'toString', 'constructor'];

	for (var i in { toString: 1 }) { namespace._enumerables = null; }

	namespace._apply = function(object, config, defaults) {
		if(defaults) { namespace._apply(object, defaults); }
		if (object && config && typeof config === 'object') {
			var i, j, k;
			for (i in config) { object[i] = config[i]; }
			if (namespace._enumerables) {
				for (j = namespace._enumerables.length; j--;) {
					k = namespace._enumerables[j];
					if (config.hasOwnProperty(k)) { object[k] = config[k]; }
				}
			}
		}
		return object;
	};

	/**
	 * Create new object typeof Recognizer.
	 * @name Recognizer
	 * @param {Object} [config] Configuration.
	 * @param {callback:initCallback} [config.onConnect] Callback function,
	 * that will be called after successful session initialization. Default: is absent.
	 * @param {callback:dataCallback} [config.onResult] Callback function,
	 * that will be called after speech recognition. Default: is absent.
	 * @param {callback:errorCallback} [config.onError] Default: is absent.
	 * @param {String} [config.uuid] Session UUID. Default: uuid.v1().
	 * @param {String} [config.apikey] API-key. Default: is absent.
	 * @param {config.format} Audio stream format. Default: PCM16.
	 * Available values:
	 * <ul>
	 *	<li>PCM16: 'audio/x-pcm;bit=16;rate=16000';</li>
	 * </ul>
	 * @param {String} [config.url] Server recognition URL.
	 * Default: 'wss://webasr.yandex.net/asrsocket.ws'.
	 * @param {Boolean} [config.punctuation] Use punctuation. Default: true.
	 * @param {Boolean} [config.allowStrongLanguage] Allow obscent vocabulary.
	 * Default: true.
	 * @param {String} [config.model] Lexical model for recognition, Default: 'notes'.
	 * Available values:
	 * <ul>
	 *	<li>'freeform'</li>
	 *	<li>'notes'</li>
	 *	<li>'queries'</li>
	 *	<li>'names'</li>
	 *	<li>'dates'</li>
	 *	<li>'maps'</li>
	 *	<li>'numbers'</li>
	 * </ul>
	 * @param {String} [config.lang] Language. Default: 'ru-RU'.
	 * Available values: 'ru-RU'; 'en-US'; 'tr-TR'; 'uk-UA'.
	 * @param {String} [config.applicationName] Application name. Default: 'jsapi'
	 * @param {Boolean} [config.partialResults] Send partial results to server.
	 * Default: true.
	 */
	var Recognizer = function(config) {

		if (!(this instanceof namespace.Recognizer)) {
			return new namespace.Recognizer(config);
		}

		this.config = namespace._apply({}, config, namespace._defaults);
		this.recognizedText = false;
		this.config.key = this.config.apikey;

	};

	Recognizer.prototype = /** @lends Recognizer.prototype */{

		/**
		* Write string `str` to object `view` at the specified offset.
		*/
		_writeStr: function(view, offset, str) {

			for (var i = 0; i < str.length; i += 1) {
				view.setUint8(offset + i, str.charCodeAt(i));
			}
		},

		/**
		* Return Int16Array with 44 bytes RIFF header.
		*/
		_getChunkBuff: function(cbLength) {

			var ab = new ArrayBuffer(44 + cbLength),
					dv = new DataView(ab), cb = new Int16Array(ab);

			this._writeStr(dv, 0, 'RIFF');				// RIFF identifier
			dv.setUint32(4, 44 + cbLength, true); // File length
			this._writeStr(dv, 8, 'WAVE');				// RIFF type
			this._writeStr(dv, 12, 'fmt ');				// Format chunk identifier
			dv.setUint32(16, 16, true);						// Format chunk length
			dv.setUint16(20, 1, true);						// Sample format (1 is PCM)
			dv.setUint16(22, 1, true);						// Channel count
			dv.setUint32(24, 16000, true);				// Sample Rate = Number of Samples per second
			dv.setUint32(28, 32000, true);				// SampleRate*BitsPerSample*Channels/8
			dv.setUint16(32, 2, true);						// BitsPerSample*Channels/8
			dv.setUint16(34, 16, true);						// Bits per sample
			this._writeStr(dv, 36, 'data');				// Data chunk identifier
			dv.setUint32(40, cbLength, true);			// Data chunk length

			return cb;
		},

		/**
		* Return RIFF header of the specified data buffer
		* (usually WAV file loaded into db).
		*/
		_getRiff: function(db) {

			return {

				identifier: db.toString('utf8', 0, 4),
				fileLength: db.readUInt32LE(4),
				type: db.toString('utf8', 8, 4),
				formatChunkIdentifier: db.toString('utf8', 12, 4),
				formatChunkLength: db.readUInt32LE(16),
				sampleFormat: db.readUInt16LE(20),
				channelCount: db.readUInt16LE(22),
				sampleRate: db.readUInt32LE(24),
				sampleRate_BPS_Channels_8: db.readUInt32LE(28),
				bitsPerSample_Channels_8: db.readUInt16LE(32),
				bitsPerSample: db.readUInt16LE(34),
				dataChunkIdentifier: db.toString('utf8', 36, 4),
				dataChunkLength: db.readUInt32LE(40)

			};
		},

		/**
		 * Connect recognition service.
		 */
		connect: function () {

			this.client = new W3CWebSocket(

				this.config.url, [], null, null, null, { fragmentOutgoingMessages: false }

			);
			this.client.binaryType = 'arraybuffer';

			this.client.onopen = function(e) {

				this.client.send(JSON.stringify({type: 'message', data: this.config}));

			}.bind(this);

			this.client.onmessage = function(e) {

				var message = JSON.parse(e.data);
				if (message.type == 'InitResponse'){

					this.config.onConnect(message.data.sessionId, message.data.code);

				} else if (message.type == 'AddDataResponse'){

					this.config.onResult(message.data);
					if(message.data.text === '') {

						if(this.recognizedText) { this.client.close(); }

					} else {

						this.recognizedText = true;
						if(message.data.close) { this.client.close(); }

					}

				} else if (message.type == 'Error'){

					console.log('Error message:\n' + message.data + '\n');

				} else { console.log('Unknown message type: ' + message.type + '\n'); }

			}.bind(this);

			this.client.onerror = function(e) { this.config.onError(e); }.bind(this);

			this.client.onclose = function(e) { this.config.onClose(e); }.bind(this);

		},

		/**
		 * Send data to recognition service.
		 * @param {Buffer} db Input file data buffer.
		 * @param {Number} cbLength Chunk buffer length in bytes.
		 */
		send: function(db, cbLength) {

			var riff = this._getRiff(db), dbTail = riff.fileLength - 44,
					dbOffset = 44, dbIndex, cbIndex, cb;

			while(dbTail > cbLength) {

				cb = this._getChunkBuff(cbLength);
				cbIndex = 22;

				for(dbIndex = dbOffset; dbIndex < dbOffset + cbLength; dbIndex += 2) {

					cb[cbIndex] = db.readInt16LE(dbIndex);
					cbIndex += 1;

				}

				this.client.send(cb);

				dbOffset += cbLength;
				dbTail -= cbLength;

			}

			cb = this._getChunkBuff(cbLength);
			cbIndex = 22;

			for(dbIndex = dbOffset; dbIndex < dbOffset + dbTail; dbIndex += 2) {

				cb[cbIndex] = db.readInt16LE(dbIndex);
				cbIndex += 1;

			}

			this.client.send(cb);

			cb = this._getChunkBuff(cbLength);		// ... silence ...
			this.client.send(cb);
		}
	};

	namespace.Recognizer = Recognizer;

}(this));
